### PR TITLE
Frontend with Framework: 2.1 Components

### LINKED ISSUE #1709: Frontend with Framework - 2.1 Components
<hr>

## Description 📜
Components- Add Proper documentation on React  Components. It should contain Introduction to React Components, Types of react components, code examples based on them.

<hr>

## Domain of Contribution 📊

- [ ] Android Dev (Flutter, Java, Kotlin)
- [ ] Backend Dev (Java, .NET, Python)
- [ ] C/CPP
- [ ] Competitive Programming
- [ ] Cyber Security
- [ ] DSA
- [ ] Database
- [ ] Datascience with Python
- [ ] Datascience with R
- [ ] Frontend Dev HTML/CSS/JS
- [×] Frontend Dev React/Angular/Vue
- [ ] Golang
- [ ] Interview Prep
- [ ] Java_Domain
- [ ] JavaScript
- [ ] MERN
- [ ] Machine Learning
- [ ] Open Source
- [ ] Python_Domain
- [ ] Rust
- [ ] Statistics
- [ ] UI/UX

<hr>

### PR DESCRIPTION
**Pull request by Utkarsh1504**
<hr>

## Description 📜
I have added proper documentation on the `Components` using the simplest example of React App so that anyone can understand.
Fixes #1709

<hr>

## Type of change 📝

- [ ] Audio (Should be in mp3 format Includes speech clarity, Concise ,Low distortion)
- [ ] Video (Animations, screen-recordings, presentations and regular explanatory films are all possibilties etc)
- [×] Documentation (Content Creation in the form of codes or tutorials)
- [ ] Other (If you choose other, Please mention changes below) 

<hr>

## Domain of Contribution 📊

- [ ] Android Dev (Flutter, Java, Kotlin)
- [ ] Backend Dev (Java, .NET, Python)
- [ ] C/CPP
- [ ] Competitive Programming
- [ ] Cyber Security
- [ ] DSA
- [ ] Database
- [ ] Datascience with Python
- [ ] Datascience with R
- [ ] Frontend Dev HTML/CSS/JS
- [×] Frontend Dev React/Angular/Vue
- [ ] Golang
- [ ] Interview Prep
- [ ] Java_Domain
- [ ] JavaScript
- [ ] MERN
- [ ] Machine Learning
- [ ] Open Source
- [ ] Python_Domain
- [ ] Rust
- [ ] Statistics
- [ ] UI/UX

<hr>

## Checklist ✅

- [×] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [×] I have performed a self-review of my own code or work.
- [×] I have commented my code, particularly in hard-to-understand areas.
- [×] My changes generates no new warnings.
- [×] I'm GWOC'21 contributor

<hr>

<hr>

### CODE REVIEW

File: Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+- There are two types of components, `function` components and `class` components. In this example we used a function component. We'll see class components shortly.
+- The `return` statement is the only thing that is required in a function component.
+- then we render our component `App` to the DOM using `ReactDOM.render`.
+- `ReactDOM.render` is what takes our rendered App component and puts in the DOM (in our case we're putting it in the root element.)
+
+## Types of Components
+
+There are two types of components in React.
+1. `function` components - these are the ones that are written in javascript functions. A function component must return markup.
+example:
```
Comment (riasenpai) [inline Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:35]: it accepts some arguments called props and returns an React element- include this too please and elaborate a bit. This seems a little incomplete.

Comment (Utkarsh1504) [reply Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:35]: Okay I'll do it.

**Comment by Utkarsh1504** [comment]
@anandhu720 @riasenpai I have added a piece of proper and complete information for React Components covering props also.
please review the changes. Thank you!

**Review by the-pro — COMMENTED** [review]
Please also include to wrap up the jsx components into a single element.

**Comment by Utkarsh1504** [comment]
@the-pro please elaborate a little bit more?

**Comment by the-pro** [comment]
> @the-pro please elaborate a little bit more?

I was saying mention the fact that we need to return a single html element in react components. All things wrapped up inside it

**Comment by Utkarsh1504** [comment]
@the-pro ok I got it.

File: Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+const App = () => {
+    return (
+        <h1>Hello World!</h1>
+    );
+};
+
+ReactDOM.render(<App />, document.getElementById("root"));
+```
+This is a simplest React Application anyone can write, let's understand it line by line:
+- The first thing we do here, is to make our own component, `App`. React is all about making components. And then taking those components and making more components out of those.
```
Comment (mitraditya) [inline Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:25]: We don't make components out of every single component. We just do it for the ones where reusability or modularity is possible. Please add this.

File: Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+The above one is also a function component, which is written in pure Javascipt function.
+after that we can render it to the DOM using `ReactDOM.render`. This is the best practise for writing React Components, most of the React developer follow this practise. 
+
+The output will be:
+
+```html
+Hello World!
+```
+
+2. `class` components - class components are written in ES6 classes. Every class component extends `React.Component`. Every class component must have a render method that returns some sort of JSX / markup / call to `React.createElement`.
```
Comment (mitraditya) [inline Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:98]: The statement "Every class component extends React.Component" is incorrect. We can make our own class and inherit from that class, to be able to use the methods defined in that parent class.

File: Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+> Now the above code is not going run for now because components needs to be export and imported. which is explained in the next lesson 2.4 `import, export, modules`. After that you will able to run this code.
+
+And you will output like this:
+```yml
+Hello John! Welcome to GWOC!
+Hello Jimmy! Welcome to GWOC!
+Hello Jack! Welcome to GWOC!
+```
+This is a complete Information about React Components. Now you should practise by your hand to create some components and pass some props to it and try to render it, you will be able to understand the concept of components. Okay let's disscuss the types of it.
```
Comment (mitraditya) [inline Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:67]: *discuss

**Review by mitraditya — CHANGES_REQUESTED** [review]
Kindly make the suggested changes.

**Comment by Utkarsh1504** [comment]
@mitraditya I have resolved all the things which you have suggested.
please review if there is anything that needs to change, kindly let me know

File: Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md [hunk 5]
```diff
@@ -95,7 +95,7 @@ The output will be:
 Hello World!
 ```
 
-2. `class` components - class components are written in ES6 classes. Every class component extends `React.Component`. Every class component must have a render method that returns some sort of JSX / markup / call to `React.createElement`.
+2. `class` components - class components are written in ES6 classes. We can make our own class and inherit from that class, to be able to use the methods defined in that parent class. Every class component must have a render method that returns some sort of JSX / markup / call to `React.createElement`.
```
Comment (mitraditya) [inline Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:98]: In general applications, we inherit from React.Component. But, we can also make our own class and inherit from that class, to be able to use the methods defined in that parent class.

File: Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md [hunk 6]
```diff
@@ -22,7 +22,7 @@ const App = () => {
 ReactDOM.render(<App />, document.getElementById("root"));
 ```
 This is a simplest React Application anyone can write, let's understand it line by line:
-- The first thing we do here, is to make our own component, `App`. React is all about making components. And then taking those components and making more components out of those.
+- The first thing we do here, is to make our own component, `App`. React is all about making components. And then taking those components to reuse and modularize in our application.
```
Comment (mitraditya) [inline Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md:25]: And then taking those components to implement reusability and modularity in our application.

**Review by mitraditya — COMMENTED** [review]
Please make these changes.

**Comment by mitraditya** [comment]
Looks fine. Just make the change with @the-pro suggested, and we're good to go.

**Comment by Utkarsh1504** [comment]
@the-pro I have included wrapping up the jsx components into a single element. please review my changes.

**Review by the-pro — COMMENTED** [review]
Nicely done!

**Pull request merged**